### PR TITLE
[Elastic Agent] Fix leaking connections and timers

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

This PR update `elastic-agent-client` dependency which fixes some issues in how grpc connections are closed and other small leak issues.

Also updates patterns of time.After usage in elastic-agent.
Also closes connection explicitly and guarding connections with context.WithCancel where it makes sense

## Why is it important?

Without this we're eating too much memory over time.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

TODO:
figure out tests to catch these kind of issues

### CODE REVIEW

File: x-pack/elastic-agent/pkg/agent/application/periodic.go [hunk 1]
```diff
@@ -31,10 +31,12 @@ func (p *periodic) Start() error {
 
 	WORK:
 		for {
+			t := time.NewTimer(p.period)
 			select {
 			case <-p.done:
+				t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/agent/application/periodic.go:37]: I think using `defer t.Stop()` below the `t := time.NewTimer(p.period)` is more future proof and more golang like. That way if a new case is ever added to the select then `t.Stop()` will already be handled.

File: x-pack/elastic-agent/pkg/agent/application/upgrade/crash_checker.go [hunk 2]
```diff
@@ -66,10 +66,13 @@ func (ch *CrashChecker) Run(ctx context.Context) {
 	ch.log.Debug("Crash checker started")
 	for {
 		ch.log.Debugf("watcher having PID: %d", os.Getpid())
+		t := time.NewTimer(ch.checkPeriod)
+
 		select {
 		case <-ctx.Done():
+			t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/agent/application/upgrade/crash_checker.go:73]: Same as above.

File: x-pack/elastic-agent/pkg/agent/application/upgrade/error_checker.go [hunk 3]
```diff
@@ -49,10 +49,12 @@ func NewErrorChecker(ch chan error, log *logger.Logger) (*ErrorChecker, error) {
 func (ch *ErrorChecker) Run(ctx context.Context) {
 	ch.log.Debug("Error checker started")
 	for {
+		t := time.NewTimer(statusCheckPeriod)
 		select {
 		case <-ctx.Done():
+			t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/agent/application/upgrade/error_checker.go:55]: Same as above.

File: x-pack/elastic-agent/pkg/composable/controller.go [hunk 4]
```diff
@@ -138,14 +139,15 @@ func (c *controller) Run(ctx context.Context, cb VarsCallback) error {
...
+				case <-t.C:
 					exitloop = true
-					break
 				}
 				if exitloop {
 					break
 				}
 			}
+
+			t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/composable/controller.go:150]: Same as above.

File: x-pack/elastic-agent/pkg/composable/providers/host/host.go [hunk 5]
```diff
@@ -52,11 +52,12 @@ func (c *contextProvider) Run(comm corecomp.ContextProviderComm) error {
 	// Update context when any host information changes.
 	go func() {
 		for {
+			t := time.NewTimer(c.CheckInterval)
 			select {
 			case <-comm.Done():
+				t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/composable/providers/host/host.go:58]: Same as above.

File: x-pack/elastic-agent/pkg/core/plugin/process/app.go [hunk 6]
```diff
@@ -296,9 +296,11 @@ func gracefulKill(proc *process.Info) {
 	wg.Wait()
 
 	// kill in case it's still running after timeout
+	t := time.NewTimer(procExitTimeout)
 	select {
 	case <-doneChan:
-	case <-time.After(procExitTimeout):
+		t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/core/plugin/process/app.go:302]: Same as above.

File: x-pack/elastic-agent/pkg/core/retry/retrystrategy.go [hunk 7]
```diff
@@ -51,10 +52,12 @@ func Do(ctx context.Context, config *Config, fn func(ctx context.Context) error,
 		}
 
 		if retryNo < retryCount {
+			t := time.NewTimer(getDelayDuration(config, retryNo))
 			select {
-			case <-time.After(getDelayDuration(config, retryNo)):
+			case <-t.C:
 			case <-ctx.Done():
-				break
+				t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/core/retry/retrystrategy.go:59]: Same as above.

File: x-pack/elastic-agent/pkg/core/server/server.go [hunk 8]
```diff
@@ -861,10 +861,12 @@ func (as *ApplicationState) destroyCheckinStream() {
 func (s *Server) watchdog() {
 	defer s.watchdogWG.Done()
 	for {
+		t := time.NewTimer(s.watchdogCheckInterval)
 		select {
 		case <-s.watchdogDone:
+			t.Stop()
```
Comment (blakerouse) [inline x-pack/elastic-agent/pkg/core/server/server.go:867]: Same as above.

**Review by blakerouse — APPROVED** [review]
Change looks good. Highlighted a nit-pick with `t.Stop()`. I think overall changing to that pattern would be better, but I will leave it up to you.

File: x-pack/elastic-agent/pkg/agent/application/upgrade/crash_checker.go (hunk 2 above)
Comment (michalpristas) [reply x-pack/elastic-agent/pkg/agent/application/upgrade/crash_checker.go:73]: we cant defer in an endless loop

File: x-pack/elastic-agent/pkg/agent/application/periodic.go (hunk 1 above)
Comment (michalpristas) [reply x-pack/elastic-agent/pkg/agent/application/periodic.go:37]: i tried to use defer whenever possible but i avoided this in a for loop.
as a follwup i can do func breakdown as in elastic-agent-client and extract loop body in a func which will be cleaner and allow using of defer statements
in this PR as this is blocker for 7.14 i want change to be as little intrusive as possible

File: x-pack/elastic-agent/pkg/agent/application/pipeline/actions/handlers/handler_action_policy_change.go [hunk 9]
```diff
@@ -143,7 +144,10 @@ func (h *PolicyChange) handleFleetServerHosts(ctx context.Context, c *config.Con
 	}
 	ctx, cancel := context.WithTimeout(ctx, apiStatusTimeout)
 	defer cancel()
-	_, err = client.Send(ctx, "GET", "/api/status", nil, nil, nil)
+	resp, err := client.Send(ctx, "GET", "/api/status", nil, nil, nil)
+	// discard body for proper cancellation and connection reuse
+	io.Copy(ioutil.Discard, resp.Body)
```
Comment (andrewkroh) [inline x-pack/elastic-agent/pkg/agent/application/pipeline/actions/handlers/handler_action_policy_change.go:149]: On error I assume `resp` is nil.

**Comment by michalpristas** [comment]
e2e seems unrelated and same across multiple tests

**Pull request merged**